src/concentratedLiquidity/provision.ts

```diff
@@ -62,45 +44,172 @@ export abstract class PositionProvider {
         contractAddress: string,
         batchDetails: BatchLPDetails,
         txOptions?: TransactionOptions,
+        marginAccountAddress?: string,
+        assetsDeposit?: Record<string, { amount: BigNumber; decimal: number }>,
     ): Promise<ethers.providers.TransactionRequest> {
         const address = await signer.getAddress();
 
+        if (!marginAccountAddress || !assetsDeposit) {
+            // no state overrides; rely on provided txOptions.gasLimit
+            const { prices, flipPrices, sizes, isBuy } = PositionProvider.buildBatchInputs(batchDetails);
+
+            const orderbookInterface = new ethers.utils.Interface(orderbookAbi.abi);
+            const data = orderbookInterface.encodeFunctionData('batchProvisionLiquidity', [
+                prices,
+                flipPrices,
+                sizes,
+                isBuy,
+                false,
+            ]);
+
+            return buildTransactionRequest({
+                from: address,
+                to: contractAddress,
+                signer,
+                data,
+                txOptions,
+            });
+        }
+
+        const { gasLimit, data } = await PositionProvider.estimateGas(
+            signer,
+            contractAddress,
+            batchDetails,
+            marginAccountAddress,
+            assetsDeposit,
+        );
+
+        const mergedTxOptions = txOptions && txOptions.gasLimit !== undefined ? txOptions : { ...txOptions, gasLimit };
+
+        return buildTransactionRequest({
+            from: address,
+            to: contractAddress,
+            signer,
+            data,
```

Comment:
data is only build in the if block, the data has to be built outside to use it here.